### ISSUE TITLE
[GetGVKsFromAddToSchem] Identify better ways to handling metrics filter for third-party API's and/or CRD which are cluster scoped.

### ISSUE THREAD

**Issue by akoserwal** [operator-framework/operator-sdk#2798]
## Feature Request

[user-guide: default-metrics-exported-with-3rd-party-resource](https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#default-metrics-exported-with-3rd-party-resource)

Current documentation has gaps in understanding about the implementation to handle third party API schemas and their dependent schemas not registered with the manager. Also, I suspect there is an issue with registered schemes that also failed to be filtered for the OCP environment.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
Suggested solution [#2794 
2794](https://github.com/operator-framework/operator-sdk/pull/2794)

Added variadic string param to GetGVKsFromAddToScheme for customkinds
it will allow to filter the custom third party or ocp schemas by passing
a list of kinds.

**Purpose: Identify a solution that can make it easier either by improving the documentation or  implementation.**

**Comment by joelanford** [comment]
@akoserwal For the sake of completeness (and for others having similar problems), can you describe the problem you're having?

> The kubemetrics.GenerateAndServeCRMetrics function requires an RBAC rule to list all GroupVersionKinds in the list of watched namespaces, so you might need to filter the kinds returned by k8sutil.GetGVKsFromAddToScheme more stringently to avoid authorization errors such as Failed to list *unstructured.Unstructured. You may also need to add a rule to LIST your third party API schemas and their dependent schemas not registered with the manager.

I do think this section is slightly confusing. `GenerateAndServeCRMetrics` creates metrics about instances of each of the GVKs passed in, in the set of namespaces passed in. It's been awhile since I've personally looked at this, but IIRC, kube-state-metrics will do a `List()` for each of these GVKs in each of those namespaces. Therefore the operator needs an RBAC rule to list that GVK in each of the specified namespaces.

The original use case for ` k8sutil.GetGVKsFromAddToScheme` was to use it only for _your_ operator's APIs, so it would return just the GVKs of the CRDs that your operator owns. Your RBAC roles would already have rules that permit `list` on these resources so this generally "just works".

So let me ask some more specific questions:
1. As a sanity check, can you explain which GVKs you're trying to create metrics for? It sounds like you're going for your operator's APIs AND some external APIs?
2. If you are specifically intending to generate metrics for the external APIs, how are you deciding which of those you want metrics for? If it is a specific subset, have you considered just appending that subset to the list returned by `k8sutil.GetGVKsFromAddToScheme` when you call it for your operator's APIs?

**Comment by joelanford** [comment]
Oh and lastly, what does your `role.yaml` file look like? Have you added `list` rules for each of the GVKs that is being passed into `GenerateAndServeCRMetrics()`?

**Comment by camilamacedo86** [comment]
HI @akoserwal, 

After spoke with @joelanford about it I also agree that  #2794 would not be the best approach.

So, may the possible solution here is to improve the docs as joe describe above and then we can : 

- Clarify how to customize the filter in the docs by providing some example such as https://github.com/3scale/3scale-operator/blob/master/cmd/manager/main.go#L240
- Add the info regards the OCP schemas caveats. See that when we add the OCP schema to the manager it also can add its dependencies schemas and then, if the role file has just the permissions to the schema added the issue can still be faced because of kubemetrics requires permissions to List and Watch all GKVs in all ns informed as well. Example: If we add `imagev1 "github.com/openshift/api/image/v1"` then others schemas will be added by default. See: https://github.com/openshift/api/blob/c99157bc6492218400b18cde7f66510f9ee56a9a/image/v1/register.go#L16

I hope that it helps.

**Comment by akoserwal** [comment]
Operator: https://github.com/teiid/teiid-operator

Issues I got when I try to update the service monitoring and metrics code generated by the operator-sdk(v15 or v16)

Errors:
``` Failed to list *unstructured.Unstructured: replicationcontrollers.extensions is forbidden:
      Failed to list *unstructured.Unstructured: deploymentconfigs.apps.openshift.io "rollback" not found
      Failed to list *unstructured.Unstructured: nodes is forbidden: 
     Failed to list *unstructured.Unstructured: imagesignatures.image.openshift.io is forbidden
     Failed to list *unstructured.Unstructured: pods "proxy"
     Failed to list *unstructured.Unstructured: deploymentconfigs.apps.openshift.io "log" not found
 ```
Many more similar to above for different resources.

[role.yaml](https://github.com/teiid/teiid-operator/blob/master/deploy/role.yaml)

**Comment by joelanford** [comment]
@akoserwal For the types in those error messages, are you intending to generate metrics for them? Or are you trying to figure out how to tell operator-sdk NOT to generate metrics for them so that those errors stop?

Where in your code are you calling `GenerateAndServeCRMetrics()`?

**Comment by akoserwal** [comment]
we revered the change: https://github.com/teiid/teiid-operator/commit/eb86d2f7d9c8df62b4b240e8aa8a895c515d4200

This case for now: "NOT to generate metrics for them so that those errors stop?"

**Comment by akoserwal** [comment]
Errors stopped after the grant all permissions to the resources, not just with verbs [list, watch].

**Comment by raffaelespazzoli** [comment]
I'm seeing this error too. And I can confirm that is not permission related as it can be reproduced even with cluster-admin role.

**Comment by camilamacedo86** [comment]
HI @raffaelespazzoli  and @akoserwal, 

Note that some OCP APIs has NOT the Watch and List verbs that are required for kubemetrics be able to List and to Watch the resources to generate the metrics from them in the specified namespaces.

**Example**

See that the ImageStreamTags is not implementing the watch which would be important to observe changes in this resource as well.
```
$ oc api-resources -o wide | grep ImageStream
imagestreamimages isimage image.openshift.io true ImageStreamImage [get]
imagestreamimports image.openshift.io true ImageStreamImport [create]
imagestreammappings image.openshift.io true ImageStreamMapping [create]
imagestreams is image.openshift.io true ImageStream [create delete deletecollection get list patch update watch]
imagestreamtags istag image.openshift.io
```

So the GKV of `ImageStreamTags` can NOT be in the filter arg passed to the `GetGVKsFromAddToScheme`. And see that you cannot grant permission for an action that is not available in the OCP API. 

And then, please beware of the purpose of this method as described previously here by @joe 

> The original use case for k8sutil.GetGVKsFromAddToScheme **was to use it only for your operator's APIs**, so it would return just the GVKs of the CRDs that your operator owns. Your RBAC roles would already have rules that permit list on these resources so this generally "just works".

So, the question is would really make sense you pass ALL these OCP values to the kubemetrics implementation? I do not think so. See that because of this, would be recommended you filter them as it is done in the 3-scale operator for example. See: https://github.com/openshift/api/blob/c99157bc6492218400b18cde7f66510f9ee56a9a/image/v1/register.go#L16

**Comment by raffaelespazzoli** [comment]
@camilamacedo86 I admit I am not fully following the discussion here. But I am not trying to do anything special with `GetGVKsFromAddToScheme`. I am writing a normal operator that uses a few OCP resources. That's all I do and I get all of those errors. When I remove metrics, I stop receiving those errors. Btw, I am not using resources from the `image.openshift.io` api domain. I'm using apis from the following domains:
`network.openshift.io`
`cloudcredential.openshift.io`
`config.openshift.io`

**Comment by akoserwal** [comment]
@camilamacedo86: This makes sense now.

Points of confusion

1) About the setting the role's permissions correctly and passing ALL these OCP values to the kubemetrics implementation?

- "kubemetrics.GenerateAndServeCRMetrics function requires an RBAC rule to list all GroupVersionKinds in the list of watched namespaces, ": 

This implies that we need to add the verb "list" to the resource defined in the role.yaml. We can add points that can make it clear to verify the correct permissions for the resources and you need to add custom filter implementation to the existing GKVs defined in the used runtime.Schema for the operator. Reason: "kube-metrics tries to list all of the defined Kinds in the schemas that are passed," including Kinds that the operator doesn't use and thus the role used the operator. We can filter the roles which are not required.

with some example like:
https://github.com/3scale/3scale-operator/blob/master/cmd/manager/main.go#L240

**Comment by camilamacedo86** [comment]
HI @raffaelespazzoli, 

> @camilamacedo86 I admit I am not fully following the discussion here. But I am not trying to do anything special with GetGVKsFromAddToScheme. I am writing a normal operator that uses a few OCP resources. That's all I do and I get all of those errors. When I remove metrics, I stop receiving those errors. 

Note that: 
- This scenario can just/only be faced when third-party APIs schemas are added on the projects
- The usage of third-party API's is not mandatory for Operator's projects works in the specific cluster vendors.

So, what I am trying to clarify here is that:

- IMO the default metrics helper provided will attend perfectly well the most common scenarios and users as their needs. 

However, if a developer would like to take advantage of OCP APIs and/or any other third-party one which is specific for some kind of vendor or some specific cluster configuration then, some customizations in the project will be required as described in the docs (which I agree as well that has a good hall for improvements). 

Also, note that other recommendations would be suggested as well for these very specific solutions and scenarios. I'd like to share with you the blogs [Why not couple an Operator's logic to a specific Kubernetes platform?](https://developers.redhat.com/blog/2020/01/22/why-not-couple-an-operators-logic-to-a-specific-kubernetes-platform/) and [How to use third-party APIs in Operator SDK projects](https://developers.redhat.com/blog/2020/02/04/how-to-use-third-party-apis-in-operator-sdk-projects/)

> Btw, I am not using resources from the image.openshift.io api domain. I'm using APIs from the following domains:
> network.openshift.io
> cloudcredential.openshift.io
> config.openshift.io

That was just an **example** to illustrate the scenario. 

Also, note that OCP Schemas may add other ones which are its dependencies. E.g. If we add `“github.com/openshift/api/image/v1”` then others schemas will be added behind of the scene. See its implementation [here](https://github.com/openshift/api/blob/c99157bc6492218400b18cde7f66510f9ee56a9a/image/v1/register.go#L16)

And then, IMO the appropriate questions for your specific scenario would be that one done by @joelanford above: **For the types in those error messages, are you intending to generate metrics for them?**  Instead of going there and grant permissions.

**Comment by joelanford** [comment]
The other key here is that `GetGVKsFromAddToScheme` is a convenience helper function that makes it easier to get your operator's GVKs from the existing registration functions.

However, if those registration functions are pulling in a bunch of other GVKs you don't care about, it is totally reasonable to skip using `GetGVKsFromAddToScheme()` and manually construct the list of GVKs to pass them into `GenerateAndServeCRMetrics()`.

**Comment by raffaelespazzoli** [comment]
@camilamacedo86 I am still confused on what is the right thing to do to make the error go away. As I said I import these APIs: 
```
network.openshift.io
cloudcredential.openshift.io
config.openshift.io
```
for my operator exactly with the method described by this article: [How to use third-party APIs in Operator SDK projects](https://developers.redhat.com/blog/2020/02/04/how-to-use-third-party-apis-in-operator-sdk-projects/). Yet I'm getting this error. 
I'm not getting from this conversation what one is supposed to do to create an operator correctly.

**Comment by camilamacedo86** [comment]
HI @raffaelespazzoli, 

As described above  

> The original use case for k8sutil.GetGVKsFromAddToScheme was to use it only for your operator's APIs, so it would return just the GVKs of the CRDs that your operator owns. Your RBAC roles would already have rules that permit list on these resources so this generally "just works".

Then, the suggested solution is to customize your metrics implementation as described in the [doc](https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#default-metrics-exported-with-3rd-party-resource). 

Note that `filteredGVK` should have just the Operator's GKV's in your scenario.  And then, as shared some example over how you can customize the implementation of the metric in your project is: https://github.com/3scale/3scale-operator/blob/master/cmd/manager/main.go#L240

**Comment by raffaelespazzoli** [comment]
@camilamacedo86 I think I get it now. here is what my code looks like:
```
	gvkFilters := []schema.GroupVersionKind{

		// Operator resource types
		schema.GroupVersionKind{Group: "redhatcop.redhat.io", Kind: "EgressIPAM", Version: matchAnyValue},
	}
```
just reporting a small fragment. The rest is the same as the one in the provided 3scale example. 
Yet I am still getting the error...

**Comment by camilamacedo86** [comment]
HI @raffaelespazzoli, 

Note that you should pass just the operator's GKVs to the `GenerateAndServeCRMetrics`. Also, you need to ensure that your operator has the required permissions to watch/list them. So, if this information is not enough for you sorted it out, I'd kindly ask for you raise a new issue providing all information as the error faced, your main.go implementation, roles and as the steps used to test/check it. 

PS.: Note that the goal of this issue is to ` Identify better ways to handling metrics filter for third-party API's and/or CRD which are cluster scoped.` which shows that check and help with specific scenarios would be out of context. 

Really thank you in advanced for your attention and comprehension.

**Comment by joelanford** [comment]
We captured this issue as an action item to address as part of #2811, so I'm going to close this one since there will likely be larger upcoming changes after the kubebuilder integration lands.

There's been a bunch of discussion here, but I don't think we've properly communicated the origin of this function and some of the assumptions we made when it was added and included in the `main.go` scaffold. So here goes!

---

When running `operator-sdk new` to create a new project, `main.go` gets scaffolded. From the start, it has this `GetGVKsFromAddToScheme()` function included. Despite there not being any APIs added to the project, this works! How you ask? 

As part of `operator-sdk new`, we also scaffold `pkg/apis/apis.go` which includes a global `AddToSchemes` SchemeBuilder and a global `AddToScheme` function. In a brand new project, nothing has been appended to `AddToSchemes`, so `GetGVKsFromAddToScheme()` returns an empty slice of GVKs.

Next, after running `operator-sdk add api --api-version=example.com/v1 --kind=App`, the `pkg/apis/example/` directory gets scaffolded with the API types AND `pkg/apis/addtoscheme_example_v1.go` gets scaffolded. That file appends the new API to the global `AddToSchemes` SchemeBuilder.

Now, without any changes in `main.go`, `GetGVKsFromAddToScheme` returns the new GVK. As more APIs are added to the project, this function ensures they are automatically configured for kube-state-metrics instrumentation.

What we didn't foresee was users customizing `pkg/apis/addtoscheme_example_v1.go` to append extra AddToScheme functions from external CRDs. Instead, we believed these external APIs would be added to the manager's scheme directly in `main.go`, as documented [here](https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#register-with-the-managers-scheme).

**TL;DR**

**Our assumptions about how `GetGVKsFromAddToScheme()` would be used don't hold. I think the correct solution for those facing this problem is to completely stop using `GetGVKsFromAddToScheme()` and to instead manually manage the set of GVKs passed into  `GenerateAndServeCRMetrics()`**

**Comment by camilamacedo86** [comment]
Just to supplement as described above note that the original use case for [`k8sutil.GetGVKsFromAddToScheme`](https://godoc.org/github.com/operator-framework/operator-sdk/pkg/k8sutil#GetGVKsFromAddToScheme) was to return just the GVKs of the CRDs that the operator owns. Your RBAC roles would have rules that permit List on these resources already. However, following an example over how would be possible to filter the 3rd party resources schemas added to the Manager. 

**Example**

```go
func filterGKVsFromAddToScheme() ([]schema.GroupVersionKind, error) {
    filteredGVK, err := k8sutil.GetGVKsFromAddToScheme(apis.AddToScheme)
    if err != nil {
        return nil, err
    }
	
    // types that will be ignored
    ignoreList := []schema.GroupVersionKind{
        // OCP types
        schema.GroupVersionKind{Group: "route.openshift.io", Kind: "Route", Version: "*"},
        schema.GroupVersionKind{Group: "image.openshift.io", Kind: "ImageStream", Version: "*"},
        schema.GroupVersionKind{Group: "apps.openshift.io", Kind: "DeploymentConfig", Version: "*"},
        // Others Operators API
        schema.GroupVersionKind{Group: "example.appservice.com", Kind: "AppService", Version: "*"},
        // Kubernetes types 
        schema.GroupVersionKind{Kind: "PersistentVolumeClaim", Version: "*"},
        schema.GroupVersionKind{Kind: "ServiceAccount", Version: "*"},
    }

    ownGVKs := []schema.GroupVersionKind{}
    for _, gvk:= range filteredGVK {
        for _, ignoreGKV := range ignoreList {
            if !hasMatch(ignoreGKV, gvk) {
            	ownGVKs = append(ownGVKs, gvk)
            }
        }
    }

    return ownGVKs, nil
}

func hasMatch(ignoreGKV schema.GroupVersionKind, gvk schema.GroupVersionKind) bool {
    if ignoreGKV.Kind == "*" && ignoreGKV.Group == "*" && ignoreGKV.Version == "*" {
        return false
    } else {
        if (ignoreGKV.Kind != "*" && ignoreGKV.Kind != gvk.Kind) ||
            (ignoreGKV.Group != "*" && ignoreGKV.Group != gvk.Group) ||
            (ignoreGKV.Version != "*" && ignoreGKV.Version != gvk.Version) {
            return false
        }
    }
}
```
